libcontainer/factory_linux.go

```diff
@@ -239,19 +239,23 @@ func (l *LinuxFactory) StartInitialization() (err error) {
 		}
 		*pair.v = i
 	}
+
 	var (
 		pipe = os.NewFile(uintptr(pipefd), "pipe")
 		it   = initType(os.Getenv("_LIBCONTAINER_INITTYPE"))
 	)
+
+	// ensure that this pipe is always closed in the case of failure.
+	defer pipe.Close()
```

Comment:
While I'm aware of the semantics of `defer`, my point is that you have to remember the semantics while reading the `defer`d functions (which also interact with errors). I mean, ultimately I don't care all that much it's just more stuff to keep track of in an already confusing function.